### PR TITLE
Fixes to Opportunist and Mirror Herb adjustments

### PR DESCRIPTION
**Pull request by AlexOn1ine**
Fixes:
``` 
Opportunist stacking boosts in one turn from opposing mons
Increasing the stat by the wrong amount when copying opposing boost (happened for +6 boosts
There is no message
Play animation only once for every stat boost
Mirror Herb does not have any animation when copying stat besides the held item anim
```

### CODE REVIEW

**Comment by AlexOn1ine** [comment]
Ready for review

File: src/battle_script_commands.c [hunk 1]
```diff
@@ -17000,3 +17006,34 @@ void BS_TryQuash(void)
...
+                SET_STATCHANGER(stat + 1, gQueuedStatBoosts[battler].statChanges[stat], FALSE);
+
+            gQueuedStatBoosts[battler].stats &= ~(1 << stat);
+            gBattlerTarget = battler;
+            gBattlescriptCurrInstr = cmd->nextInstr;
+            return;
+        }
+    }
+    gBattlescriptCurrInstr = cmd->jumpInstr;
+}
```
Comment (Bassoonian) [inline src/battle_script_commands.c:17039]: missing newline

File: data/battle_scripts_1.s [hunk 2]
```diff
@@ -9151,16 +9151,23 @@ BattleScript_MirrorHerbCopyStatChange::
 	printstring STRINGID_MIRRORHERBCOPIED
 	waitmessage B_WAIT_TIME_LONG
 	removeitem BS_SCRIPTING
-	call BattleScript_TotemVar_Ret
-	copybyte gBattlerAttacker, sSAVED_BATTLER	@ restore the original attacker just to be safe
+BattleScript_MirrorHerbStartCopyStats:
```
Comment (Skeli789) [inline data/battle_scripts_1.s:9154]: As discussed on Discord, this actually should function the same way as Opportunist.

File: src/battle_script_commands.c [hunk 3]
```diff
@@ -11781,10 +11781,16 @@ static u32 ChangeStatBuffs(s8 statValue, u32 statId, u32 flags, const u8 *BS_ptr
...
+            if ((statValue + gBattleMons[battler].statStages[statId]) > MAX_STAT_STAGE)
+                statIncrease = MAX_STAT_STAGE - gBattleMons[battler].statStages[statId];
+            else
+                statIncrease = statValue;
+
             gBattleCommunication[MULTISTRING_CHOOSER] = (gBattlerTarget == battler);
             gProtectStructs[battler].statRaised = TRUE;
 
-            // check mirror herb
+            // Check Mirror Herb / Opportuniest
```
Comment (Skeli789) [inline src/battle_script_commands.c:11793]:
```suggestion
            // Check Mirror Herb / Opportunist
```

File: src/battle_script_commands.c [hunk 4]
```diff
@@ -17000,3 +17006,34 @@ void BS_TryQuash(void)
...
+    NATIVE_ARGS(u8 battler, const u8 *jumpInstr);
+    u32 stat = 0;
+    u32 battler = GetBattlerForBattleScript(cmd->battler);
+
+    if (gQueuedStatBoosts[battler].stats == 0)
+    {
+        memset(gQueuedStatBoosts, 0, sizeof(gQueuedStatBoosts)); // erase all boosts so they are not stacked
+        gBattlescriptCurrInstr = cmd->jumpInstr;    // stats done, exit
+        return;
+    }
```
Comment (Skeli789) [inline src/battle_script_commands.c:17021]: Won't this break Opportunist and Mirror Herb that activate at the same time?

File: src/battle_script_commands.c [hunk 5]
```diff
@@ -17000,3 +17006,34 @@ void BS_TryQuash(void)
...
+
+    for (stat = 0; stat < (NUM_BATTLE_STATS - 1); stat++)
```
Comment (Skeli789) [inline src/battle_script_commands.c:17023]: This goes through the order Speed -> Sp. Atk -> Sp. Def, no? Probably should fix it so it goes through the correct visual "order"

File: src/battle_util.c [hunk 6]
```diff
@@ -5837,9 +5837,24 @@ u32 AbilityBattleEffects(u32 caseID, u32 battler, u32 ability, u32 special, u32
...
+                        if ((gQueuedStatBoosts[battler].stats & (1 << i)) == 0)
+                            continue;
+
+                        if (!statBuffMoreThan1)
+                            statBuffMoreThan1 = ((gQueuedStatBoosts[battler].stats & (1 << i)) > 1);
+
+                        if (gBattleScripting.animArg1 != 0) //Already set in a different stat so now boosting multiple stats
+                            gBattleScripting.animArg1 = (!statBuffMoreThan1 ? STAT_ANIM_MULTIPLE_PLUS1 : STAT_ANIM_MULTIPLE_PLUS2);
+                        else
+                            gBattleScripting.animArg1 = GET_STAT_BUFF_ID(i) + (!statBuffMoreThan1 ? STAT_ANIM_PLUS1 : STAT_ANIM_PLUS2);
```
Comment (Skeli789) [inline src/battle_util.c:5855]: Test to make sure it doesn't need to be
```suggestion
                            gBattleScripting.animArg1 = GET_STAT_BUFF_ID(i + 1) + (!statBuffMoreThan1 ? STAT_ANIM_PLUS1 : STAT_ANIM_PLUS2);
```

**Review by Skeli789 — CHANGES_REQUESTED** [review]

File: src/battle_util.c (hunk 6 above)
Comment (AlexOn1ine) [reply src/battle_util.c]: yep, it was needed

File: src/battle_script_commands.c (hunk 5 above)
Comment (AlexOn1ine) [reply src/battle_script_commands.c:17035]: So I'm not sure if this is optimal but I delay the speed check so it is always last

File: src/battle_script_commands.c (hunk 4 above)
Comment (AlexOn1ine) [reply src/battle_script_commands.c]: yes, there was actually a bug regardless if I erased it or not. I tried to work around that by setting all the bits back that have a stat increase.

**Comment by AlexOn1ine** [comment]
Ready for review again. 

I had to port the changes from #4971 to `TryDoEventsBeforeFirstTurn` and would like if this wasn't hold up by 1.10 because it is a crucial bug fix. I needed it to skip totem boosts after opportunist was triggered once. It was possible for the totem boost to go off. Also erasing the totem boosts happened to late.

File: include/battle_main.h [hunk 7]
```diff
@@ -23,6 +23,26 @@ struct MultiPartnerMenuPokemon
...
+    FIRST_TURN_EVENTS_TERRAIN,
+    FIRST_TURN_EVENTS_STARTING_STATUS,
+    FIRST_TURN_EVENTS_TOTEM_BOOST,
+    FIRST_TURN_EVENTS_NEUTRALIZING_GAS,
+    FIRST_TURN_EVENTS_SWITCH_IN_ABILITIES,
+    FIRST_TURN_EVENTS_OPPORTUNIST_1,
+    FIRST_TURN_EVENTS_ITEM_EFFECTS,
+    FIRST_TURN_EVENTS_OPPORTUNIST_2,
+    FIRST_TURN_EVENTS_END,
+};
```
Comment (Bassoonian) [inline include/battle_main.h:44]: If we're hoping to have this still be in master before 1.9, can we perhaps not use this to-be-figured-out enum system but stick with good old defines for now? We can obviously refactor this later, but it feels a bit out of place in the current ecosystem

Comment (AlexOn1ine) [reply include/battle_main.h]: turned it into a simple enum

File: test/battle/hold_effect/mirror_herb.c [hunk 8]
```diff
@@ -35,7 +35,7 @@ SINGLE_BATTLE_TEST("Mirror Herb copies all of foe's positive stat changes in a t
     }
 }
 
-SINGLE_BATTLE_TEST("Mirror Herb copies all of Stuff Cheeks' stat boosts")
+SINGLE_BATTLE_TEST("Mirror Herb copies all of Stuff Cheeks stat boosts")
```
[the remark this answers is not in the document]
Comment (AlexOn1ine) [reply test/battle/hold_effect/mirror_herb.c]: reverted

**Review by Bassoonian — APPROVED** [review]

File: include/battle_main.h [hunk 9]
```diff
@@ -23,6 +23,20 @@ struct MultiPartnerMenuPokemon
 #define BOUNCE_MON          0x0
 #define BOUNCE_HEALTHBOX    0x1
 
+enum {
```
Comment (Skeli789) [inline include/battle_main.h:26]: Style: `{` should be on new line

File: src/battle_util.c [hunk 10]
```diff
@@ -5874,9 +5874,40 @@ u32 AbilityBattleEffects(u32 caseID, u32 battler, u32 ability, u32 special, u32
...
                 {
-                    gBattleScripting.savedBattler = gBattlerAttacker;
-                    gBattleScripting.battler = gBattlerAttacker = gBattlerAbility = battler;
+                    bool32 statBuffMoreThan1 = FALSE;
+                    bool32 handleSpeedAnimLater = FALSE;
+                    gBattleScripting.animArg1 = 0;
+                    gBattleScripting.battler = battler;
                     gProtectStructs[battler].activateOpportunist--;
+
+                    for (i = 0; i < (NUM_BATTLE_STATS - 1); i++)
```
Comment (Skeli789) [inline src/battle_util.c:5883]: Ideally you'd have some constant array declaring a "visual stat order" and i is the index in that array which contains stat ids. That way you don't need a `handleSpeedAnimLater` variable.

File: src/battle_util.c [hunk 11]
```diff
@@ -6815,10 +6846,43 @@ static u8 TryConsumeMirrorHerb(u32 battler, bool32 execute)
...
+        bool32 handleSpeedAnimLater = FALSE;
+        gBattleScripting.animArg1 = 0;
+
         gLastUsedItem = gBattleMons[battler].item;
-        gBattleScripting.savedBattler = gBattlerAttacker;
-        gBattleScripting.battler = gBattlerAttacker = battler;
+        gBattleScripting.battler = battler;
         gProtectStructs[battler].eatMirrorHerb = 0;
+
+        for (i = 0; i < (NUM_BATTLE_STATS - 1); i++)
```
Comment (Skeli789) [inline src/battle_util.c:6858]: This is a lot of duplicate code you can make a function to handle for both.

**Pull request merged**